src/Winton.Extensions.Configuration.Consul/ConsulConfigurationProvider.cs

```diff
@@ -2,74 +2,86 @@
 
 using System;
-using System.Linq;
+using System.Net;
+using System.Threading;
 using System.Threading.Tasks;
 using Consul;
 using Microsoft.Extensions.Configuration;
-using Microsoft.Extensions.Primitives;
 using Winton.Extensions.Configuration.Consul.Extensions;
 
 namespace Winton.Extensions.Configuration.Consul
 {
-    internal sealed class ConsulConfigurationProvider : ConfigurationProvider
+    /// <summary>
+    ///     Each instance loads configuration for the key in Consul that is specified in
+    ///     the contained <see cref="IConsulConfigurationSource"/>.
+    ///     It has the ability to automatically reload the config if it changes in Consul.
+    /// </summary>
+    /// <remarks>
+    ///     Each instance maintains its own <c>lastIndex</c> and uses this to detect changes.
+    ///     Each instance ensures calls to Consul are serialised, to avoid concurrent access to <c>lastIndex</c>.
+    /// </remarks>
+    internal sealed class ConsulConfigurationProvider : ConfigurationProvider, IDisposable
     {
-        private readonly IConsulConfigurationClient _consulConfigClient;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly IConsulClientFactory _consulClientFactory;
         private readonly IConsulConfigurationSource _source;
+        private ulong _lastIndex;
+        private Task _pollTask;
 
         public ConsulConfigurationProvider(
             IConsulConfigurationSource source,
-            IConsulConfigurationClient consulConfigClient)
+            IConsulClientFactory consulClientFactory)
         {
             if (source.Parser == null)
             {
                 throw new ArgumentNullException(nameof(source.Parser));
             }
 
-            _consulConfigClient = consulConfigClient;
             _source = source;
+            _consulClientFactory = consulClientFactory;
+            _cancellationTokenSource = new CancellationTokenSource();
+        }
 
-            if (source.ReloadOnChange)
-            {
-                ChangeToken.OnChange(
-                    () => _consulConfigClient.Watch(_source.Key, _source.OnWatchException, _source.CancellationToken),
-                    async () =>
-                    {
-                        await DoLoad(true).ConfigureAwait(false);
-                        OnReload();
-                    });
-            }
+        public void Dispose()
+        {
+            _cancellationTokenSource.Cancel();
+            _pollTask?.Wait(500);
+            _cancellationTokenSource.Dispose();
         }
 
         public override void Load()
         {
-            DoLoad(false).GetAwaiter().GetResult();
+            // If polling has already begun then calling load is pointless
+            if (_pollTask != null)
+            {
+                return;
+            }
+
+            DoLoad().GetAwaiter().GetResult();
```

Comment:
`.GetAwaiter().GetResult()` is preferred as it will unwrap an aggregate exception in the same way that `await` would. Generally it provides neater and more readable stack traces in the case of errors. It also can simplify the path to `async` by meaning that existing catch blocks will continue working the same way.